### ISSUE TITLE
Why can't this library generate runtime types?

### ISSUE THREAD

**Issue by benwinding** [gcanti/io-ts#243]
Sorry, this is probably a dumb question, but why can't this library _generate_ the "runtime type" from the native typescript interfaces and types at compile time?

Instead of having to manually redefine all types in the alternative _io-ts_ type definitions style.

**Comment by smpalileo** [comment]
Not sure if I understand your question correctly but typescript is still javascript and it has no runtime types, so io-ts just wraps them in a class or something like that.

**Comment by benwinding** [comment]
Yes Typescript is _still_ Javascript I get that, but Typescript is compiled using the `tsc` command right?

So, why can't io-ts generate it's **run-time types**  when Typescript is compiled with the `tsc` command? Instead of redefining, using the cubersome `io-ts` system. 

## For example:
### Before tsc

``` typescript
interface Person {
  name: string;
  age: number;
}
```

### After tsc
``` typescript
const Person = t.type({
  name: t.string,
  age: t.number
})
```

Then we can easily add this library to existing projects with hundreds of interfaces already defined

Is this not possible?

**Comment by giogonzo** [comment]
@benwinding unless I'm missing something (compiler plugins maybe) there's no way to have `tsc` produce such code at compile time. You would need an external tool in your pipeline.

This library uses the opposite approach, using `io-ts` types as the source of truth, from which static/plain TS types can be derived with `t.TypeOf<typeof iotsType>`. It is possible to derive static types from runtime types, but not the other way around (tsc never produces any runtime code, except for a few older features like `enum`s)

**Comment by smpalileo** [comment]
I think there's a gap in compile time where interfaces declared in typescript disappear because they don't exist in javascript. You'll need to run a function that does something like this:

 case (_.isString(value)):
          return t.string;
        case (_.isInteger(value)):
          return t.Integer;
        case (_.isNumber(value)):
          return t.number;
        case (_.isBoolean(value)):
          return t.boolean;

I use lodash to get the types btw

**Comment by smpalileo** [comment]
I tried to  look everywhere for that kind of tool but I just couldn't find any that automatically does that. The closest one I found was a JSON schema to interface converter and even that didn't work as well as I need it to.

**Comment by MastroLindus** [comment]
I also couldn't find it, and that's the reason I implemented io-ts types in my code. 
But I think it should be possible, and if it would exist, I would definitely prefer to have typescript as the source of truth than an arbitrary type-system created by a single library.
While it's not there, io-ts is definitely a fantastic second choice though.

**Comment by mmkal** [comment]
There's a library that attempts to do something like what you're describing, but it doesn't seem to be maintained: https://github.com/fabiandev/ts-runtime

**Comment by andykais** [comment]
one such actively maintained project does exist: https://github.com/woutervh-/typescript-is though this library only does validation. E.g. wrapping a nullable field in an fp-ts option is out of the scope of the project.
